src/consensus/progress_map.zig

```diff
@@ -0,0 +1,1169 @@
+/// the actual required synchronization semantics are to be determined later.
+const stubs = struct {
+    fn Arc(comptime T: type) type {
+        return struct { arc_ed: T };
+    }
+    fn RwLock(comptime T: type) type {
+        return struct { rwlock_ed: T };
+    }
+};
+
+// TODO: move this somewhere better?
+/// NOTE: usage of this type indicates that the integer
+/// must only be operated on using saturating arithmetic.
+/// This is used to annotate types as saturating, to match
+/// agave's type definitions.
+fn Saturating(comptime T: type) type {
+    return enum(T) {
+        _,
+
+        pub fn toInt(self: Saturating(T)) T {
+            return @intFromEnum(self);
+        }
+
+        /// Returns `self`, casted to a pointer to `T`, retaining all attributes.
+        pub fn asInt(self: anytype) AsInt(@TypeOf(self)) {
+            return @ptrCast(self);
+        }
+
+        pub fn AsInt(comptime Self: type) type {
+            const lazy = struct {
+                fn nonMatchErr() noreturn {
+                    @compileError("Expected single item pointer to " ++ @typeName(Saturating(T)));
+                }
+            };
+
+            var info = switch (@typeInfo(Self)) {
+                .Pointer => |info| info,
+                .Optional => |info| switch (@typeInfo(info.child)) {
+                    .Pointer => return ?AsInt(info.child),
+                    else => lazy.nonMatchErr(),
+                },
+                else => lazy.nonMatchErr(),
+            };
+            if (info.size != .One) lazy.nonMatchErr();
+            if (info.child != Saturating(T)) lazy.nonMatchErr();
+            info.child = T;
+
+            return @Type(.{ .Pointer = info });
+        }
+    };
+}
+
+/// AUDIT: https://github.com/anza-xyz/agave/blob/cb32984a9b0d5c2c6f7775bed39b66d3a22e3c46/core/src/consensus/progress_map.rs
+pub const ProgressMap = struct {
+    map: std.AutoArrayHashMapUnmanaged(Slot, ForkProgress),
+
+    pub const INIT: ProgressMap = .{ .map = .{} };
+
+    pub fn deinit(self: ProgressMap, allocator: std.mem.Allocator) void {
+        var map = self.map;
+        for (map.values()) |fork_progress| fork_progress.deinit(allocator);
+        map.deinit(allocator);
+    }
+
+    pub fn clone(
+        self: ProgressMap,
+        allocator: std.mem.Allocator,
+    ) std.mem.Allocator.Error!ProgressMap {
+        var result: ProgressMap = INIT;
+        errdefer result.deinit(allocator);
+
+        try result.map.ensureTotalCapacity(allocator, self.map.count());
+        for (self.map.keys(), self.map.values()) |k, v| {
+            result.map.putAssumeCapacityNoClobber(k, try v.clone(allocator));
+        }
+
+        return result;
+    }
+
+    pub fn getForkStats(self: ProgressMap, slot: Slot) ?*ForkStats {
+        const fork_progress = self.map.getPtr(slot) orelse return null;
+        return &fork_progress.fork_stats;
+    }
+
+    pub fn getHash(self: *const ProgressMap, slot: Slot) ?Hash {
```

Comment:
This and getForkStats are the methods that are directly used in the tower. Although the underlying map would be populated somewhere else (during the replay?) having these two methods is enough to plug it in, into the tower.

Reply:
`progress.my_latest_landed_vote` is another function off the `ProgressMap` that has popped up in the new sets of tower related methods I am adding.

Reply:
...and just ran into a usage of [get_propagated_stats](https://github.com/anza-xyz/agave/blob/8e5470a9c4c01f5725e2d37c475abf6980fcb435/core/src/consensus/progress_map.rs#L280) also

Reply:
and also [get_leader_propagation_slot_must_exist](https://github.com/anza-xyz/agave/blob/8e5470a9c4c01f5725e2d37c475abf6980fcb435/core/src/consensus/progress_map.rs#L347)

Reply:
I think I would like to add these in a follow up PR, the scope of this PR has reached a boiling point with other methods recently added requested by Drew.